### PR TITLE
Fix broken ARM64 NEON encodings and add a WORD-encoding validator

### PR DESCRIPTION
**Pull request by tphakala**
## Summary

Adds a static validator for hand-encoded ARM64 `WORD $0x...` directives and fixes three shipping NEON bugs it helped surface.

The validator (`internal/asmencoding` + `asmcheck_test.go`) decodes every `WORD` directive with `golang.org/x/arch/arm64asm` and asserts it matches the instruction named in its comment. It only decodes bytes, so it runs under `go test ./...` on any architecture and guards CI without ARM hardware. It would have caught the historical FNEG/FRINTN (#6) and REV64 encoding regressions.

Bugs fixed (all hardware-confirmed on a Raspberry Pi 5, all hidden by tests that used fewer elements than the SIMD vector width, so the NEON path never ran):

- **f16 `Reciprocal`** returned 0 for any length >= 8. `0x4E032042` decoded to `TBL`, not `INS V2.D[1],V3.D[0]`, which zeroed the `1.0h` numerator. Fixed to `0x6E180462`.
- **f16 `Min` / `Max`** ignored every element past the first 8 lanes. `0x4EC03480` / `0x4E403480` wrote the result to `V0` instead of the `V4` accumulator. Fixed to `0x4EC03484` / `0x4E403484`.
- **f32 FMINP / FMAXP** reductions encoded `Vm=V16` instead of `V0` at four sites (`0x6EB0F400` -> `0x6EA0F400`, `0x6E30F400` -> `0x6E20F400`). Harmless today because the polluted lanes are discarded, but the encoding read an unintended register.

FP16 `.8H` instructions cannot be decoded by `arm64asm`; the 19 distinct ones are allow-listed, each verified once with a native aarch64 `objdump`.

## Related Issues

- Relates to #44 (statically validate FP16 encodings via objdump cross-check)
- Relates to #45 (audit SIMD test coverage for vector-width threshold gaps)

## Test Plan

- [x] `go test ./...` green on x86 (1810 tests)
- [x] `go test ./...` green on real ARM64 (rpi5, Go 1.26.1)
- [x] New f16 vector-path tests fail on the old encodings and pass on the fixed ones (verified on hardware)
- [x] `asmcheck` reports zero findings after the fixes; `golangci-lint` and `go vet` clean

## Summary by CodeRabbit

* **Tests**
  * Added comprehensive test coverage for ARM64 vector operation paths, including reciprocal, minimum, and maximum functions with edge cases.
  * Added validation tests for hand-encoded ARM64 assembly directives.

* **Chores**
  * Updated dependency: added `golang.org/x/arch v0.27.0`.

* **Optimizations**
  * Optimized ARM64 NEON instructions for floating-point minimum, maximum, and reciprocal operations across FP16 and FP32 implementations.

[Review Change Stack](https://app.coderabbit.ai/change-stack/tphakala/simd/pull/46?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

File: asmcheck_test.go [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+
+func allowed(file string, hex uint32, used map[int]bool) bool {
+	for i, e := range asmAllowlist {
+		if e.hex == hex && strings.HasSuffix(file, e.file) {
+			used[i] = true
+			return true
+		}
+	}
+	return false
+}
```
[the remark this answers is not in the document]
Comment (tphakala) [reply asmcheck_test.go:84]: Fixed in 9e10a27. I normalize at the walk boundary instead: findArm64Asm now stores filepath.ToSlash(p), so the value passed to allowed() and the error messages are forward-slash on every platform. The windows-latest job was the only failure and is now addressed.

File: internal/asmencoding/asmencoding.go [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+var wordLineRe = regexp.MustCompile(`WORD\s+\$0x([0-9A-Fa-f]{1,8})\b`)
+var commentRe = regexp.MustCompile(`//\s*(.*)$`)
+
+// ParseWordLine extracts the hex value and inline comment from a single
+// assembly line. ok is false when the line is not a WORD $0x... directive.
+func ParseWordLine(line string) (hex uint32, comment string, ok bool) {
+	m := wordLineRe.FindStringSubmatch(line)
+	if m == nil {
+		return 0, "", false
+	}
```
[the remark this answers is not in the document]
Comment (tphakala) [reply internal/asmencoding/asmencoding.go:110]: Fixed in 9e10a27. ParseWordLine now returns ok=false when WORD appears only after a // on the line, so a commented-out directive is ignored. Added a test case.

File: internal/asmencoding/asmencoding.go [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+		d := Directive{Line: i + 1, Hex: hex}
+		switch {
+		case inline != "":
+			d.Comment = inline
+			d.Source = InlineComment
+		case i > 0:
+			if prev := strings.TrimSpace(lines[i-1]); strings.HasPrefix(prev, "//") {
+				d.Comment = strings.TrimSpace(strings.TrimPrefix(prev, "//"))
+				d.Source = PrecedingComment
+			}
```
[the remark this answers is not in the document]
Comment (tphakala) [reply internal/asmencoding/asmencoding.go:73]: Fixed in 9e10a27. ScanSource now skips a preceding comment whose text starts with WORD, so a commented-out directive above an active one is not used as its description.

File: internal/asmencoding/asmencoding.go [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+	}
+	return arm64asm.GNUSyntax(inst), nil
+}
+
+// Normalize canonicalizes an instruction string for comparison: upper case,
+// single-spaced, no space after commas, with aliased mnemonics folded.
+func Normalize(s string) string {
+	s = strings.ToUpper(strings.TrimSpace(s))
+	s = strings.Join(strings.Fields(s), " ")
+	s = strings.ReplaceAll(s, ", ", ",")
```
[the remark this answers is not in the document]
Comment (tphakala) [reply internal/asmencoding/asmencoding.go:138]: Fixed in 9e10a27. Normalize now strips spaces before commas as well, with a test.

File: internal/asmencoding/asmencoding.go [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+			d.Source = InlineComment
+		case i > 0:
+			if prev := strings.TrimSpace(lines[i-1]); strings.HasPrefix(prev, "//") {
+				// A commented-out WORD directly above is disabled code, not a
+				// description of this instruction.
+				if text := strings.TrimSpace(strings.TrimPrefix(prev, "//")); !strings.HasPrefix(text, "WORD") {
+					d.Comment = text
+					d.Source = PrecedingComment
+				}
+			}
```
[the remark this answers is not in the document]
Comment (tphakala) [reply internal/asmencoding/asmencoding.go:73]: Fixed in 1573010. ScanSource now requires the preceding comment to be non-empty and checks the WORD prefix case-insensitively, so an empty // line or a lowercase commented-out directive above an instruction is not used as its description. Added a test for the empty-comment case.
